pkg/providers/tinkerbell/tinkerbell.go

```diff
@@ -448,6 +461,102 @@ func (p *tinkerbellProvider) setMachinesToPXEBoot(ctx context.Context) error {
 	return errorutil.NewAggregate(errs)
 }
 
+// scrubWorkflowsFromTinkerbell removes all workflows in the Tinkerbell stack that feature in hardware by retrieving
+// hardware MAC addresses using tinkerbellHardware. tinkerbellHardware is necessary because MAC addresses aren't
+// available on the Hardware object type.
+func (p *tinkerbellProvider) scrubWorkflowsFromTinkerbell(ctx context.Context, hardware []tinkv1alpha1.Hardware, tinkerbellHardware []*tinkhardware.Hardware) error {
+	workflows, err := p.providerTinkClient.GetWorkflow(ctx)
+	if err != nil {
+		return fmt.Errorf("retrieving workflows: %w", err)
+	}
+
+	hardwareMACLookup, err := createHardwareIDToMACMapping(tinkerbellHardware)
+	if err != nil {
+		return err
+	}
+
+	manifestHardwareMACs, err := createMACSetFromHardwareManifests(hardwareMACLookup, hardware)
+	if err != nil {
+		return err
+	}
+
+	workflowIDs, err := getWorkflowsIDsFromMACs(manifestHardwareMACs, workflows)
+	if err != nil {
+		return err
+	}
+
+	if err := p.providerTinkClient.DeleteWorkflow(ctx, workflowIDs...); err != nil {
+		return fmt.Errorf("could not delete tinkerbell workflow: %v", err)
+	}
+
+	return nil
+}
+
+func createHardwareIDToMACMapping(hardware []*tinkhardware.Hardware) (map[string]string, error) {
+	hardwareMACLookup := make(map[string]string)
+	for _, h := range hardware {
+		if len(h.Network.Interfaces) == 0 {
+			return nil, fmt.Errorf("hardware manifest without interface: hardware ID = '%v'", h.Id)
+		}
+		hardwareMACLookup[h.Id] = h.Network.Interfaces[0].Dhcp.Mac
+	}
+
+	return hardwareMACLookup, nil
+}
+
+func createMACSetFromHardwareManifests(hardwareMACLookup map[string]string, hardware []tinkv1alpha1.Hardware) (macAddressSet, error) {
+	manifestHardwareMACs := make(macAddressSet)
+	for _, h := range hardware {
+		mac, found := hardwareMACLookup[h.Spec.ID]
+		if !found {
+			return nil, fmt.Errorf("couldn't find mac address for hardware manifest: manifest hardware ID = '%v'", h.Spec.ID)
+		}
+
+		manifestHardwareMACs.Insert(mac)
```

Comment:
Hmmm, looking more, why create this set at all? Why can't the map's key just be the mac? You are going through each hardware, creating a map of id:mac, then going through the hardware again, looking it up in the map to get the mac, then creating a set of the mac addresses, then looking up the mac of the workflow in the set using Contains... phew!